staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -17,330 +17,247 @@ limitations under the License.
-		spec:       spec,
-		handler:    specInfo.handler,
-		etag:       etag,
-	})
+// UpdateAPIService updates the api service. It is thread safe.
+func (s *specAggregator) UpdateAPIService(apiServiceName string) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return s.updateServiceLocked(apiServiceName)
 }
 
-// AddUpdateAPIService adds or updates the api service. It is thread safe.
-func (s *specAggregator) AddUpdateAPIService(handler http.Handler, apiService *v1.APIService) error {
-	s.rwMutex.Lock()
-	defer s.rwMutex.Unlock()
-
+// AddAPIService adds the api service. It is thread safe. If the
+// apiservice already exists, it will be updated.
+func (s *specAggregator) AddAPIService(apiService *v1.APIService, handler http.Handler) error {
 	if apiService.Spec.Service == nil {
-		// All local specs should be already aggregated using local delegate chain
 		return nil
 	}
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
 
-	newSpec := &openAPISpecInfo{
-		apiService: *apiService,
-		handler:    handler,
-	}
-	if specInfo, existingService := s.openAPISpecs[apiService.Name]; existingService {
-		newSpec.etag = specInfo.etag
-		newSpec.spec = specInfo.spec
+	_, exists := s.openAPISpecs[apiService.Name]
+	if !exists {
+		s.openAPISpecs[apiService.Name] = &openAPISpecInfo{
+			apiService: *apiService,
+			downloader: newNameDecoratedErrorCache(apiService.Name, NewCacheableDownloader(s.downloader, handler)),
+		}
+		s.openAPIVersionedService.UpdateSpecLazy(s.buildMergeSpecLocked())
 	}
-	return s.tryUpdatingServiceSpecs(newSpec)
+
+	return s.updateServiceLocked(apiService.Name)
 }
 
-// RemoveAPIServiceSpec removes an api service from OpenAPI aggregation. If it does not exist, no error is returned.
+// RemoveAPIService removes an api service from OpenAPI aggregation. If it does not exist, no error is returned.
 // It is thread safe.
-func (s *specAggregator) RemoveAPIServiceSpec(apiServiceName string) error {
-	s.rwMutex.Lock()
-	defer s.rwMutex.Unlock()
+func (s *specAggregator) RemoveAPIService(apiServiceName string) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
 
-	if _, existingService := s.openAPISpecs[apiServiceName]; !existingService {
-		return nil
+	if _, exists := s.openAPISpecs[apiServiceName]; !exists {
+		return ErrNotFound
 	}
-
-	return s.tryDeleteServiceSpecs(apiServiceName)
+	delete(s.openAPISpecs, apiServiceName)
+	// Re-create the mergeSpec for the new list of apiservices
+	s.openAPIVersionedService.UpdateSpecLazy(s.buildMergeSpecLocked())
+	return nil
 }
 
-// GetAPIServiceSpec returns api service spec info
-func (s *specAggregator) GetAPIServiceInfo(apiServiceName string) (handler http.Handler, etag string, exists bool) {
-	s.rwMutex.RLock()
-	defer s.rwMutex.RUnlock()
+// newNameDecoratedErrorCache creates a new cache that wraps a downloader
+// cache the name of the apiservice to help with debugging.
+func newNameDecoratedErrorCache(name string, cache cached.Data[*spec.Swagger]) cached.Data[*spec.Swagger] {
```

Comment:
```suggestion
func decoratedError(name string, cache cached.Data[*spec.Swagger]) cached.Data[*spec.Swagger] {
```
There is no other function like that around. Keep names sane.